src/MongoDB.Driver/LibMongoCrypt/EncryptionSource.cs

```diff
@@ -0,0 +1,163 @@
+﻿/* Copyright 2019-present MongoDB Inc.
+*
+* Licensed under the Apache License, Version 2.0 (the "License");
+* you may not use this file except in compliance with the License.
+* You may obtain a copy of the License at
+*
+* http://www.apache.org/licenses/LICENSE-2.0
+*
+* Unless required by applicable law or agreed to in writing, software
+* distributed under the License is distributed on an "AS IS" BASIS,
+* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+* See the License for the specific language governing permissions and
+* limitations under the License.
+*/
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using MongoDB.Bson;
+using MongoDB.Bson.IO;
+using MongoDB.Crypt;
+using MongoDB.Driver.Core.Misc;
+
+namespace MongoDB.Driver.LibMongoCrypt
+{
+    /// <summary>
+    /// Represents a encryption source.
+    /// </summary>
+    public interface IEncryptionSource //: IDisposable
+    {
+#pragma warning disable 3003
+        /// <summary>
+        /// Gets the singleton instance of <see cref="CryptClient"/>.
+        /// </summary>
+        /// <returns>The singleton instance of <see cref="CryptClient"/>.</returns>
+        CryptClient CryptClient { get; }
+#pragma warning restore
+        
+        /// <summary>
+        /// Gets the mongocryptd client.
+        /// </summary>
+        IMongoClient MongoCryptDClient { get; }
+    }
+
+    internal class EncryptionSource : IEncryptionSource
+    {
+        private readonly AutoEncryptionOptions _autoEncryptionOptions;
+        private readonly Lazy<CryptClient> _cryptClient;
+        private readonly IMongoClient _mongoCryptDClient;
+
+        private EncryptionSource(MongoClientSettings mongoClientSettings)
+        {
+            Ensure.IsNotNull(mongoClientSettings, nameof(mongoClientSettings));
+            _autoEncryptionOptions = Ensure.IsNotNull(mongoClientSettings.AutoEncryptionOptions, nameof(mongoClientSettings.AutoEncryptionOptions));
+            _mongoCryptDClient = CreateMongoCryptDClient(_autoEncryptionOptions.ExtraOptions);
+            _cryptClient = new Lazy<CryptClient>(() => CreateCryptClient(CreateCryptOptions()));
+        }
+
+        public IMongoClient MongoCryptDClient => _mongoCryptDClient;
+
+#pragma warning disable 3003
+        public CryptClient CryptClient => _cryptClient.Value;
+#pragma warning restore
+
+        public static IEncryptionSource CreateEncryptionSourceIfNecessary(MongoClientSettings mongoClientSettings)
+        {
+            if (mongoClientSettings.AutoEncryptionOptions != null)
+            {
+                return new EncryptionSource(mongoClientSettings);
+            }
+            else
+            {
+                return null;
+            }
+        }
+
+        private CryptOptions CreateCryptOptions()
```

Comment:
I moved this method here because we should save a whole list of `schemaMap` without filtering by anything. 
As we discussed with @jyemin, the reason is that there is a single MongoCrypt for the whole cluster and it has to be aware of all the schemas.